epochStart/metachain/rewardsV2.go

```diff
@@ -0,0 +1,496 @@
+func NewEpochStartRewardsCreatorV2(args RewardsCreatorArgsV2) (*rewardsCreatorV2, error) {
+	brc, err := NewBaseRewardsCreator(args.BaseRewardsCreatorArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	if check.IfNil(args.StakingDataProvider) {
+		return nil, epochStart.ErrNilStakingDataProvider
+	}
+	if check.IfNil(args.EconomicsDataProvider) {
+		return nil, epochStart.ErrNilEconomicsDataProvider
+	}
+	if big.NewInt(0).Cmp(args.TopUpGradientPoint) >= 0 {
+		return nil, epochStart.ErrInvalidRewardsTopUpGradientPoint
+	}
+	if args.TopUpRewardFactor < 0 || args.TopUpRewardFactor > 1 {
+		return nil, epochStart.ErrInvalidRewardsTopUpFactor
+	}
+
+	rc := &rewardsCreatorV2{
+		baseRewardsCreator:    brc,
+		economicsDataProvider: args.EconomicsDataProvider,
+		stakingDataProvider:   args.StakingDataProvider,
+		topUpRewardFactor:     args.TopUpRewardFactor,
+		topUpGradientPoint:    args.TopUpGradientPoint,
+	}
+
+	return rc, nil
+}
+
+// CreateRewardsMiniBlocks creates the rewards miniblocks according to economics data and validator info.
+// This method applies the rewards according to the economics version 2 proposal, which takes into consideration
+// stake top-up values per node
+func (rc *rewardsCreatorV2) CreateRewardsMiniBlocks(metaBlock *block.MetaBlock, validatorsInfo map[uint32][]*state.ValidatorInfo) (block.MiniBlockSlice, error) {
+	if check.IfNil(metaBlock) {
+		return nil, epochStart.ErrNilHeaderHandler
+	}
+
+	rc.mutRewardsData.Lock()
+	defer rc.mutRewardsData.Unlock()
+
+	miniBlocks := rc.initializeRewardsMiniBlocks()
+
+	err := rc.prepareRewardsData(metaBlock, validatorsInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	protRwdTx, protRwdShardId, err := rc.createProtocolSustainabilityRewardTransaction(metaBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	rewardsPerNode, dustFromRewardsPerNode := rc.computeRewardsPerNode(validatorsInfo)
+	log.Debug("arithmetic difference from dust rewards per node", "value", dustFromRewardsPerNode.String())
+
+	dust, err := rc.addValidatorRewardsToMiniBlocks(
+		validatorsInfo, metaBlock, miniBlocks, rewardsPerNode,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	dust.Add(dust, dustFromRewardsPerNode)
+	rc.adjustProtocolSustainabilityRewards(protRwdTx, dust)
+	err = rc.addProtocolRewardToMiniBlocks(protRwdTx, miniBlocks, protRwdShardId)
+	if err != nil {
+		return nil, err
+	}
+
+	return rc.finalizeMiniBlocks(miniBlocks), nil
+}
+
+// VerifyRewardsMiniBlocks verifies if received rewards miniblocks are correct
+func (rc *rewardsCreatorV2) VerifyRewardsMiniBlocks(metaBlock *block.MetaBlock, validatorsInfo map[uint32][]*state.ValidatorInfo) error {
+	if check.IfNil(metaBlock) {
+		return epochStart.ErrNilHeaderHandler
+	}
+
+	createdMiniBlocks, err := rc.CreateRewardsMiniBlocks(metaBlock, validatorsInfo)
+	if err != nil {
+		return err
+	}
+
+	return rc.verifyCreatedRewardMiniBlocksWithMetaBlock(metaBlock, createdMiniBlocks)
+}
+
+func (rc *rewardsCreatorV2) addValidatorRewardsToMiniBlocks(
```

Comment:
this func is similar to rewardsCreatorV1

Reply:
addValidatorRewardsToMiniBlocks

Reply:
the implementation is different.
v2 requires as input the rewards per node, and returns the dust